### ISSUE TITLE
dotnet watch only trigges on changes OUTSIDE VisualStudio 

### ISSUE THREAD

**Issue by oiBio** [dotnet/aspnetcore#34418]
### Describe the bug
After some more digging for #34355 i found out that hot reload works as excected when started from outside VisualStudio
Changes from VisualStudioCode oder the Texteditor trigger dotnet watch and i get the filechange in the console and my browser also reloades as expected
>watch : File changed: C:\***\Index.razor.
>watch : Hot reload of changes succeeded.

then i change the very same file from inside VisualStudio there is no file change in my console and also not hotreload...
Also when i start my project via VisualStudio and use the "apply code changes" i always get this. 

![image](https://user-images.githubusercontent.com/33127462/125949485-c76aa328-c708-45eb-9372-0545670de09f.png)

### To Reproduce
use cmd
change to project directory
run: dotnet watch run
wait till browser opens page

Modify "index.razor" with editor.exe or code and change text inside \<h1> Tag
Wait till hot reload is successful and browser refreshed.

Edit same file "index.razor" with VS and NOT get any File Changed message nor hot reload.

### Further technical details
- ASP.NET Core version: 
> net6-preview6
- Include the output of `dotnet --info`

> .NET SDK (reflecting any global.json):
>  Version:   6.0.100-preview.6.21355.2
>  Commit:    7f8e0d76c0
> 
> Runtime Environment:
>  OS Name:     Windows
>  OS Version:  10.0.18363
>  OS Platform: Windows
>  RID:         win10-x64
>  Base Path:   C:\Program Files\dotnet\sdk\6.0.100-preview.6.21355.2\
> 
> Host (useful for support):
>   Version: 6.0.0-preview.6.21352.12
>   Commit:  770d630b28
> 
> .NET SDKs installed:
>   2.2.207 [C:\Program Files\dotnet\sdk]
>   3.1.200-preview-014977 [C:\Program Files\dotnet\sdk]
>   3.1.200-preview-015002 [C:\Program Files\dotnet\sdk]
>   3.1.300-preview-015048 [C:\Program Files\dotnet\sdk]
>   3.1.300-preview-015095 [C:\Program Files\dotnet\sdk]
>   3.1.302 [C:\Program Files\dotnet\sdk]
>   3.1.400-preview-015151 [C:\Program Files\dotnet\sdk]
>   5.0.100-rc.2.20479.15 [C:\Program Files\dotnet\sdk]
>   5.0.100 [C:\Program Files\dotnet\sdk]
>   5.0.201 [C:\Program Files\dotnet\sdk]
>   5.0.301 [C:\Program Files\dotnet\sdk]
>   5.0.400-preview.21277.10 [C:\Program Files\dotnet\sdk]
>   6.0.100-preview.4.21255.9 [C:\Program Files\dotnet\sdk]
>   6.0.100-preview.6.21355.2 [C:\Program Files\dotnet\sdk]
> 
> .NET runtimes installed:
>   Microsoft.AspNetCore.All 2.1.26 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
>   Microsoft.AspNetCore.All 2.1.28 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
>   Microsoft.AspNetCore.All 2.2.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
>   Microsoft.AspNetCore.App 2.1.26 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
>   Microsoft.AspNetCore.App 2.1.28 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
>   Microsoft.AspNetCore.App 2.2.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
>   Microsoft.AspNetCore.App 3.1.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
>   Microsoft.AspNetCore.App 3.1.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
>   Microsoft.AspNetCore.App 3.1.15 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
>   Microsoft.AspNetCore.App 3.1.16 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
>   Microsoft.AspNetCore.App 5.0.4 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
>   Microsoft.AspNetCore.App 5.0.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
>   Microsoft.AspNetCore.App 5.0.7 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
>   Microsoft.AspNetCore.App 6.0.0-preview.6.21355.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
>   Microsoft.NETCore.App 2.1.28 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>   Microsoft.NETCore.App 2.2.8 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>   Microsoft.NETCore.App 3.1.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>   Microsoft.NETCore.App 3.1.8 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>   Microsoft.NETCore.App 3.1.15 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>   Microsoft.NETCore.App 3.1.16 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>   Microsoft.NETCore.App 5.0.4 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>   Microsoft.NETCore.App 5.0.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>   Microsoft.NETCore.App 5.0.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>   Microsoft.NETCore.App 6.0.0-preview.6.21352.12 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>   Microsoft.WindowsDesktop.App 3.1.6 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
>   Microsoft.WindowsDesktop.App 3.1.15 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
>   Microsoft.WindowsDesktop.App 3.1.16 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
>   Microsoft.WindowsDesktop.App 5.0.4 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
>   Microsoft.WindowsDesktop.App 5.0.6 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
>   Microsoft.WindowsDesktop.App 5.0.7 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
>   Microsoft.WindowsDesktop.App 6.0.0-preview.6.21353.1 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

- The IDE (VS / VS Code/ VS4Mac) you're running on, and its version
> Microsoft Visual Studio Professional 2022 Preview (64-bit)
Version 17.0.0 Preview 2.0

**Comment by mkArtakMSFT** [comment]
Thanks for contacting us. This is not part of the runtime feature-set. Please use the [VS Feedback](https://docs.microsoft.com/en-us/visualstudio/ide/feedback-options?view=vs-2019) to report this issue, so the team who owns the area can track the ask.

**Comment by realivanjx** [comment]
hot reload works fine visual studio 2022 preview though

**Comment by oiBio** [comment]
Not on my side. Dotnet watch does only trigger, when changes are from outside Visual Studio -> notepad.exe, visualcode.exe etc...
Even with 6.0.100-preview.7.21379.14 and Visual Studio 2022 Preview 2.1 :-(

**Comment by realivanjx** [comment]
did you send feedback to visual studio? if you have please share the ticket here. thanks.

**Comment by oiBio** [comment]
> did you send feedback to visual studio? if you have please share the ticket here. thanks.

https://developercommunity.visualstudio.com/t/dotnet-watch-only-trigges-on-changes-OUT/1494362

**Comment by mkArtakMSFT** [comment]
Thanks @oiBio. We've looked into this a bit more from the referenced VS Feedback issue and we believe this is an issue with dotnet watch. Reopening this and closing the linked AzDO issue to investigate this.

**Comment by mkArtakMSFT** [comment]
@oiBio we're having trouble reproducing this behavior. Can you think about any unique aspects of your setup, which may result in this? Do you have an option to try this on a different machine and see whether it reproes there too?

**Comment by oiBio** [comment]
It must have something to do with our repo. Hot-reload works with the default-template without problems. I can't determine what exactly it is, though, unfortunately. I have already changed various things and played around (launchSettings.json, global.json). Unfortunately without success 

I had also set up my own VM with Visual Studio 2022 and had the same behavior. But I could not find out exactly what the problem was. 

I do not know what is special about our project. Except maybe that we have been developing the project with Blazor since early 2019 and have migrated since then.

As soon as I can download the Preview3 of VS2022 today I will test it again and give feedback here.

**Comment by CrahunGit** [comment]
I've the same trouble, even in preview 7. It only detects changes outside visual studio.

**Comment by oiBio** [comment]
upgraded to preview7 but still the same issue. at the moment i cant download the preview3 from VS2022 - the installer does not offer an upgrade.

Also changes from VS 2019 16.11.0 or VS 2019 Preview 16.11.0 Preview 4 are not recognised. Only changes from outside VS

**Comment by CrahunGit** [comment]
Works fine on VS 2022 preview 3. Not in 2019.

**Comment by oiBio** [comment]
another observation I have made today: 

When I start the project under net5 then the file changes are recognized even if they are made via VS2022. 
Only in the upgraded project not.

Also: When i introduce a compile error like in Preview7 Code with VS2022

```
@{
     var invalid = statement
}
```

then the dotnet watch "sees" the file change after fixing and recompile the project after every file change from within VS2022 - after succesful compile and run filechanges only get recognized from outside VS

**Comment by pranavkm** [comment]
@oiBio any chance I could have you try your app using the latest nightly build of the installer: https://github.com/dotnet/installer#installers-and-binaries? We made some tweaks to the file watching over rc1 and rc2, so it's possible that the issue was fixed in those changes. 

If you continue seeing the issue after you update, one option might be to consider running `dotnet watch` with the `--verbose` switch - it should emit some additional diagnostics, perhaps that might help with the diagnosis.

**Comment by pranavkm** [comment]
@oiBio you're likely running in to this: https://github.com/aspnet/Announcements/issues/470 which has instructions on how to re-enable it. Could I bother you to post a comment in https://github.com/dotnet/aspnetcore/issues/34754 about this? It'd be interesting to figure out why you're seeing this message, but I would prefer not conflating it with this thread.

**Comment by oiBio** [comment]
Hi paranvka,
all our connections always run via wss and i got the error even then i changed the code to the proposed code snipped. i think the problem was some outdated nuget packages that where not updated. got many other error in further testings

i have massive problems with the rc2. like dependencies to older rc2 versions, some false compile errors like missing namespaces or types etc.. i think the rc2 does not work with my vs2022 preview version. i couldnt even start my web app. sometimes it worked when i manually compiles each project via dotnet build; sometimes it worked with "dotnet run" but not with "dotnet watch run" -> all startet via VS2022 PowerShell Console.

after about 2 hours of playing around with this i tried to run it via the normal command window. first time i was able to start with "dotnet watch run" 

long story short: i could run "dotnet watch run" via console prompt (only outside VS) and then started VS2022 and changed a file -> and dotnet watch triggered the file change no matter if the change was from VS2022, VS2019 or Notepad++.

so the issue might be solved, but i am not sure if everythink worked.. i had so many errors that disapeared and reappeared all the time i cant really tell what was happening :-/

**Comment by pranavkm** [comment]
Thanks @oiBio for giving it a try. 

>  sometimes it worked with "dotnet run" but not with "dotnet watch run" -> all startet via VS2022 PowerShell Console.

There appears to be an issue running `dotnet run` / `dotnet watch` from the VS console that is being tracked by https://developercommunity.visualstudio.com/t/running-a-net-6-preview-7-aspnet-core-app-in-the-v/1502210. I'm not sure if you're able to see the issue since it's internally filed, but their response goes like so:

> This issue is currently being investigated. Our team will get back to you if either more information is needed, a workaround is available, or the issue is resolved. In the meantime, you can fix this issue by removing the “DOTNET_ROOT” environment variable from your terminal session. Keep in mind you’ll have to run this command for every terminal instance you open. We know this is very inconvenient and we’re working on a permanent solution.

```
Powershell: run $env:DOTNET_ROOT = ""
Command Prompt: set DOTNET_ROOT=""
```

For the other issues you were seeing building the app in VS, it'd be great if you could use the [VS feedback tool](https://docs.microsoft.com/en-us/visualstudio/ide/how-to-report-a-problem-with-visual-studio?view=vs-2019) to report. It should help it get directed to the right team.

I'm going to close this as there isn't any further actionable work for dotnet-watch.

**Comment by muzzar78** [comment]
I have just upgraded a .Net 5 asp.net core project to .Net 6 preview 7 and using visual studio 2022 preview 3 I'm seeing the same behavior. I'm running dotnet watch run from outside visual studio. If I press save multiple times within Visual Studio occasionally dotnet watch run --verbose picks up the file change but more often than not it doesn't. If I make the changes in VS Code then every time I save the file the change gets picked up. I'll test again when the RC comes out.

**Comment by vijayrkn** [comment]
@muzzar78 - Can you please confirm if you are still running into this issue with today's release of.NET 6.0 rc1 bits?

**Comment by muzzar78** [comment]
Just confirming, the dotnet watch is working properly with rc1. 👍
